### PR TITLE
Migrate /mods command to helix API

### LINKED ISSUE #3971: Migrate /mods command to Helix API
As part of Twitch's announced deprecation of IRC-based commands ([see here for more info](https://discuss.dev.twitch.tv/t/deprecation-of-chat-commands-through-irc/40486), the `/mods` command needs to be migrated to use the relevant Helix API endpoint.

Helix API reference: https://dev.twitch.tv/docs/api/reference#get-moderators

Split from #3956

## Consequence
**Only** broadcasters can now list moderators in their own channel

### PR DESCRIPTION
**Pull request by cbclemmer**
Pull request checklist:

closes #3971 

- [x] `CHANGELOG.md` was updated, if applicable

# Description
Modeled after the /chatters migration PR #4088.

/mods command no longer uses the irc command and now makes a helix api request.
mod list is paginated just like chatter list.

### CODE REVIEW

**Comment by cbclemmer** [comment]
@pajlada I tried to follow the same paradigm as your changes to the /chatters api request.

File: src/controllers/commands/CommandController.cpp [hunk 1]
```diff
@@ -938,6 +938,85 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                    errorMessage += "Missing required scope. "
+                                    "Re-login with your "
+                                    "account and try again.";
+                }
+                break;
+
+                case Error::UserNotAuthorized: {
+                    errorMessage += "You must have moderator permissions to "
+                                    "use this command.";
+                }
```
Comment (Felanbird) [inline src/controllers/commands/CommandController.cpp:964]:
```suggestion
                case Error::UserNotAuthorized: {
                    errorMessage +=
                        "Due to Twitch restrictions, this command can only be "
                        "used by moderators. To see the list of mods you must "
                        "use the Twitch website.";
                }
```
Slightly reworded the error from the same case in `/vips`

**Comment by Felanbird** [comment]
This PR can get `helixTimegate` applied to it, as we know approximately what day the previous command will be deprecated.

The related code to copy can be found in the `/vips` PR https://github.com/Chatterino/chatterino2/pull/4053/files

File: src/controllers/commands/CommandController.cpp (hunk 1 above)
Comment (cbclemmer) [reply src/controllers/commands/CommandController.cpp]: Corrected language

**Comment by cbclemmer** [comment]
@Felanbird I added the timegate logic.

File: src/controllers/commands/CommandController.cpp [hunk 2]
```diff
@@ -938,6 +938,110 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                                "account and try again.";
+            }
+            break;
+
+            case Error::UserNotAuthorized: {
+                errorMessage +=
+                    "Due to Twitch restrictions, this command can only be "
+                    "used by moderators. To see the list of mods you must "
+                    "use the Twitch website.";
+            }
```
Comment (Felanbird) [inline src/controllers/commands/CommandController.cpp:965]:
```suggestion
            case Error::UserNotAuthorized: {
                errorMessage +=
                    "Due to Twitch restrictions,"  //
                    "this command can only be used by channel moderators. "
                    "To see the list of mods you must use the Twitch website.";
            }
```
Inserted the word `channel` and tricked the formatter so the message is easy to read via the code

**Review by Felanbird — CHANGES_REQUESTED** [review]
A few things I noticed:
- [ ] Clickable/colorized names have been regressed.
https://github.com/Chatterino/chatterino2/pull/3187/files was the original implementation
- [ ] Response should be sorted alphabetically

- [ ] needs a test in `tests/src/HighlightController.cpp`
example from /vips:
```
    // /vips
    // The extra parenthesis around the failure callback is because its type contains a comma
    MOCK_METHOD(
        void, getChannelVIPs,
        (QString broadcasterID,
         ResultCallback<std::vector<HelixVip>> successCallback,
         (FailureCallback<HelixListVIPsError, QString> failureCallback)),
        (override));  // /vips
```

File: src/controllers/commands/CommandController.cpp (hunk 2 above)
Comment (Wissididom) [reply src/controllers/commands/CommandController.cpp:965]: For me it looks like in the suggested change there will be a space missing after the comma (`restrictions,this command can only  be used`). I think it should be (Taking the suggestion as base)
````
                errorMessage +=
                    "Due to Twitch restrictions, "  //
                    "this command can only be used by channel moderators. "
                    "To see the list of mods you must use the Twitch website.";
```

Comment (Felanbird) [reply src/controllers/commands/CommandController.cpp:965]: > For me it looks like in the suggested change there will be a space missing after the comma 

Yes you were correct, resolved

File: src/controllers/commands/CommandController.cpp [hunk 3]
```diff
@@ -938,6 +938,110 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                        {
+                            message = message + QString(", and ");
+                        }
+                        else if (listSize == 2 && i == 1)
+                        {
+                            message = message + QString(" and ");
+                        }
+
+                        i++;
+                    }
```
[the remark this answers is not in the document]
Comment (cbclemmer) [reply src/controllers/commands/CommandController.cpp]: I moved the logic to the `TwitchMessageBuilder::listOfUsersSystemMessage` function and that did everything was done here.

**Comment by cbclemmer** [comment]
@Felanbird All items from your previous comment should be fixed.

File: src/controllers/commands/CommandController.cpp [hunk 4]
```diff
@@ -938,6 +938,97 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                    for (auto it = result.begin(); it != result.end(); it++)
+                    {
+                        list << it->userName;
+                    }
+
+                    list.sort(Qt::CaseInsensitive);
+
+                    MessageBuilder builder;
+                    TwitchMessageBuilder::listOfUsersSystemMessage(
+                        "The moderators of this channel are ", list,
```
Comment (Felanbird) [inline src/controllers/commands/CommandController.cpp:1021]:
```suggestion
                        "The moderators of this channel are", list,
```
before:
image
after:
image

**Review by Felanbird — APPROVED** [review]
Functionality works as expected 👍

File: src/singletons/Settings.hpp [hunk 5]
```diff
@@ -442,6 +442,10 @@ class Settings : public ABSettings, public ConcurrentSettings
         "/misc/twitch/helix-timegate/vips",
         HelixTimegateOverride::Timegate,
     };
+    EnumSetting<HelixTimegateOverride> helixTimegateModerators = {
+        "/misc/twitch/helix-timegate/moderaters",
```
Comment (pajlada) [inline src/singletons/Settings.hpp:446]:
```suggestion
        "/misc/twitch/helix-timegate/moderators",
```

File: src/providers/twitch/api/Helix.hpp [hunk 6]
```diff
@@ -1120,6 +1166,13 @@ class Helix final : public IHelix
         ResultCallback<HelixChatters> successCallback,
         FailureCallback<HelixGetChattersError, QString> failureCallback);
 
+    // Get moderator list - This method is what actually runs the API request
+    // https://dev.twitch.tv/docs/api/reference#get-moderators
+    void fetchModerators(
+        QString broadcasterID, QString after,
```
Comment (pajlada) [inline src/providers/twitch/api/Helix.hpp:1172]: This function should have an `int first` parameter too like fetchChatters

The usages of this should probably just set this to 100

File: src/providers/twitch/api/Helix.hpp [hunk 7]
```diff
@@ -1095,6 +1132,15 @@ class Helix final : public IHelix
         ResultCallback<HelixChatters> successCallback,
         FailureCallback<HelixGetChattersError, QString> failureCallback) final;
 
+    // Get Chatters from the `broadcasterID` channel
```
Comment (pajlada) [inline src/providers/twitch/api/Helix.hpp:1135]:
```suggestion
    // Get moderators from the `broadcasterID` channel
```

File: src/providers/twitch/api/Helix.hpp [hunk 8]
```diff
@@ -1095,6 +1132,15 @@ class Helix final : public IHelix
...
+    // This will follow the returned cursor
+    // https://dev.twitch.tv/docs/api/reference#get-moderators
+    void getModerators(
```
Comment (pajlada) [inline src/providers/twitch/api/Helix.hpp:1138]: `getModerators` should accept a `maxModeratorsToFetch` like `getChatters` does

**Review by pajlada — CHANGES_REQUESTED** [review]
Small change requests - last PR before Helix migration is done 
image

**Comment by pajlada** [comment]
> I tried to follow the same paradigm as your changes to the /chatters api request.

You followed the paradigm really well, and it turns out my implementation actually didn't work! I made it work now for sure, and I'll make a PR later to fix the Chatters command too

**Review by pajlada — APPROVED** [review]

**Pull request merged**